### ISSUE TITLE
Object directory for all build artifacts

### ISSUE THREAD

**Issue by imphil** [lowRISC/opentitan#650]
Currently, our various build script produce binaries all over the tree.
- Hardware build artifacts are in `build/`.
- Software build artifacts are in `sw/build` (for Make) and some other directories for meson.

This setup is problematic for a variety of reasons.
- It makes it hard to separate build and test pipelines into stages, and have artifact transfer between the stages. 
- It makes it hard to rely on artifacts from another build step.
  For example, the FPGA build includes the bootrom, so it needs to reference the boot_rom.vmem file in a reasonably stable way.
- In-tree builds (which is "build artifacts intermixed with source code", as we do today) make it hard to implement a "make distclean", i.e. "remove all build artifacts and start over".

Can we
- Keep this in mind when developing the meson (or any other) build system improvements?
- Create a single top-level directory for all build artifacts? (a typical name would be `$REPO_TOP/objdir`), and roughly document its structure?

I'd go for something like this, mirroring the structure in the source tree.

```
$REPO_TOP/objdir
|- sw/device/fpga (or other target name)
|- sw/device/sim (or other target name)
|- sw/host/...
|- hw/top_earlgrey_nexysvideo/...
```

Note that the object directory is *not equal to a distribution directory*; the object directory contains intermediate build outputs (like *.o files), which would not be present in a distribution directory. (A distribution directory, as created by a hypothetical "make dist" step, would take binary artifacts out of the objdir and create a binary distribution archive from it. Let's keep this issue separate for now.)

@mcy @gkelly

**Comment by imphil** [comment]
I did prepare a proposal draft at https://docs.google.com/document/d/153s9kH4SyQif88K1k-6GzLCmTj1RCa4V_eWbY-hdrD4/edit# This is mostly how I'd envision things to be; we will need to tweak the proposal to find a common denominator with how meson likes things to be organized. (But meson is not set in stone either, we can improve it as necessary to work for our use case.) @mcy can you have a look?

Please feel free to comment and modify the document if you have ideas. Possibly we can add this to the agenda of the software meeting tomorrow @gkelly?

**Comment by mcy** [comment]
I think that doc does a good job of expressing a lot of issues I currently see with how we build stuff, thanks for writing it up. I've left some individual comments one stuff that I think could cause problems going forward, but otherwise looks like a reasonable design (and somewhat approximates what I did in #917, up to change of directory names).

**Comment by imphil** [comment]
@mcy Now that you've implemented a modified form of the directory structure proposed in the Google Doc, would you mind documenting it as part of https://github.com/lowRISC/opentitan/blob/master/doc/ug/directory_structure.md?

**Comment by mcy** [comment]
@imphil Well, first I want to add the two hardware artifacts we ship, the Verilated system and the FPGA bitstream, to that structure. But, yes, we should document build-bin's structure sooner rather than later.

**Comment by lenary** [comment]
I think this has now been closed by 3c8ab3b so I'm going to close this issue.
